File: 프로그래머스/1/131705. 삼총사/삼총사.cpp
#include <string>
#include <vector>

using namespace std;

int answer = 0;

void combination(vector<int> number, int index, int curr, int total) {
    if (index == 3) {
        if (total == 0) {
            answer++;
        }
        return;
    }
    
    for (int i = curr; i < number.size(); i++) {
        total += number[i];
        combination(number, index + 1, i + 1, total);
        total -= number[i];
    }
}

int solution(vector<int> number) {
    
    combination(number, 0, 0, 0);
    
    return answer;
}

File: 프로그래머스/1/131705. 삼총사/삼총사.java
class Solution {
    
    int answer = 0;
    int[] number;
    
    public int solution(int[] number) {
        this.number = number;
        combination(0, 0, 0);
        return answer;
    }
    
    private void combination(int current, int index, int total) {
        if (index == 3) {
            if (total == 0) {
                answer++;
            }
            return;
        }
        
        for (int i = current; i < number.length; i++) {
            total += number[i];
            combination(i+1, index+1, total);
            total -= number[i];
        }
    }
}